### ISSUE TITLE
[PRINCIPLED OBJECTION] The VC data model specification 1.0 is not implementable in the ways it appears to be intended based on the text of the current specification document

### ISSUE THREAD

**Issue by mwherman2000** [w3c/vc-data-model#797]
Literally a copy of https://github.com/w3c/vc-data-model/issues/793#issuecomment-902715529 (re-created as a top-level issue)...
 
> To understand how the spec looks the way it looks today, you have to go back through the history to understand how we got here.

Thank you for the background explanation ...and the detailed history.

But I beg to differ @kdenhartog, I need and should only need to read the current text of the VC data model specification to understand it and to correctly implement software based on it. The actual text of what the VC WG is being asked to "ratify" is the _story_ the way in appears in the text of the current specification. This is the _story_ we're being asked to "approve" ...not the history that led to the current text.

[PRINCIPLED OBJECTION] The VC data model specification 1.0 is not implementable in the ways it appears to be intended to be implemented based on the text of the current specification document (based on the above discussion).
[PROPOSED RESOLUTION] In the VC data model specification, explicitly detail the bindings between the VC conceptual data model (https://www.w3.org/TR/vc-data-model/#core-data-model) and at least 3 target data platforms/formats:
- JSON. Reference: ?
- JSON-LD. Reference: https://github.com/w3c/vc-data-model/issues/793#issuecomment-902461256
- JPEG (as an example of a non-JSON format). Reference: https://github.com/w3c/vc-data-model/issues/796

CC: @brentzundel @wyc

**Comment by mwherman2000** [comment]
For example, IMO it's impossible to implement a JSON-LD variation of VCs that's interoperable with other JSON-LD implementations based on the current text of the VC specification.  For example, this kind of detail is absent from the specification but necessary to implement a conforming, interoperable JSON-LD VC platform.

**Comment by kdenhartog** [comment]
I think you may have misunderstood what I was saying there. "understand how the spec looks the way it looks today" is not the same to me as "not implementable in the ways it appears to be intended based on the text of the current specification document". 

This appears to be a conflation behind the story that hold the concepts of the VC spec together and the literal underpinnings oh what the spec is specifying. My original point was to explain _only_ about the story that hold the concepts together. This "story" is really just about trying to convey a mental model to assist the reader in understand why the normative statements have been made. Nothing would stop an implementer from choosing to not adhere with the mental model of the spec and still write a conformant implementation. However doing so does run a greater risk of software not behaving in interoperable ways.

With that in mind, it's useful to understand where the spec is coming from to better understand what conflicts may arise from the new suggestions. Some of these changes may result in non-normative text to change (aka text that tells the "story") and some may result in normative statement changes (aka MUST be a JSON object). In any case both are addressable, but from what I'm gathering from your statement here I'm not able to figure out what they are.

> The VC data model specification 1.0 is not implementable in the ways it appears to be intended to be implemented based on the text of the current specification document (based on the above discussion).

Great, this helps narrow it down for me a bit to somewhere in [section 6.2](https://w3c.github.io/vc-data-model/#json-ld) needing some changes. Can you point me to a few specific sentences in there which caused you trouble in implementing this and what you're proposed changes for them would be. Additionally any preemptive explanation as to the confusion the original statement caused you and some reasons why the new statement is more helpful would be useful as well to help us found common ground faster. Here's a [good issue](https://github.com/w3c/vc-data-model/issues/602) to look at as an example.

 >In the VC data model specification, explicitly detail the bindings between the VC conceptual data model (https://www.w3.org/TR/vc-data-model/#core-data-model) and at least 3 target data platforms/formats
 
 2 of these have been documented already. See ### [section 6.1](https://w3c.github.io/vc-data-model/#json) for JSON and section 6.2 for JSON-LD. In the case of the third we'd need to define that in a normative way which remains out of scope for the working group at this point. If you'd like to add the JPEG format I'd hold off until the V2 working group to see what others would like to do with that format. The alternative is to start implementing and authoring that portion of text, but know now that there's no guarantees that this text will be added to the spec.

**Comment by mwherman2000** [comment]
@kdenhartog It's a bit late here. I've read your reply once and will reply tomorrow.  Quickly, the phrase that did jump out at me was:

> Nothing would stop an implementer from choosing to not adhere with the mental model of the spec and still write a _conformant implementation_. However doing so does run a greater risk of software not behaving in _interoperable_ ways.

_Italics_ are mine.

What is the measures/definition/criteria for being _conformant_? IMO, I believe this is kind of detail (to be detailed) that is missing and hence, prevents a developer reading the current text of implementing a conformant, interoperable specification (as, for example, I'm trying to do with the Microsoft Trinity platform using no underlying JSON or JSON-LD technologies (except for last minute serialization if JSON is needed).

An aside: I'm going to create a _macro_ issue to simple list the specific issues that call into this category. I need someway to track and think about them as a group. An issue is a good public way to maintain the list ...if folks don't mind to much.

**Comment by kdenhartog** [comment]
> What is the measures/definition/criteria for being conformant?

An implementation that's passing the [test suite](https://w3c.github.io/vc-test-suite/) implies a conformant implementation.

>An aside: I'm going to create a macro issue to simple list the specific issues that call into this category. I need someway to track and think about them as a group. An issue is a good public way to maintain the list ...if folks don't mind to much.

I'd generally prefer that issues that are relevant to you and your work would be kept separate from the WG issues. I'd suggest forking the repo and opening issues on that fork. Then as you have more specific issues around spec text changes you can file them here for the group to discuss. Often times issues that cover too many topics end up in a mess to untangle and bring to completion.

**Comment by mwherman2000** [comment]
> > What is the measures/definition/criteria for being conformant?
> 
> An implementation that's passing the [test suite](https://w3c.github.io/vc-test-suite/) implies a conformant implementation.

1. At a minimum, why aren't at least the Basic Document requirements listed in the specification?

2. The specification barely references the Test Suite at all and then only in passing at the highest level. As a developer, I would expect to see at a minimum a Conformance section in the spec that highlights the Test Suite and, again, at a minimum, lists the Basic Document characteristics.
> The Working Group has received implementation feedback showing that there are at least two implementations for each normative feature in the specification. The group has obtained reports from nine (9) implementations. For details, see the test suite and implementation report.

3. There also needs to be a lot more clarity around in all of the documentation that "VC Specification" is not a single document but rather a collection or family of documents. Perhaps, there should be a central document that is actually called the "Verifiable Credentials Specification" ...2 to 5 pages long that highlights this.

"More news at 11..."

**Comment by kdenhartog** [comment]
> * At a minimum, why aren't at least the Basic Document requirements listed in the specification?
Not sure what you mean by "basic document" here. Could you clarify or do you have an example of where another spec is doing something similar?

> * The specification barely references the Test Suite at all and then only in passing at the highest level. As a developer, I would expect to see at a minimum a Conformance section in the spec that highlights the Test Suite and, again, at a minimum, lists the Basic Document characteristics.

The purpose of the specification is to specify, so that seems like it would be a conflation. Instead this is provided by way of the [charter](https://www.w3.org/2017/vc/WG/) where all of the related repositories to the WG are made available.

>There also needs to be a lot more clarity around in all of the documentation that the "VC Specification" is not a single document but rather a collection or family of documents. Perhaps, there should be a central document that is actually called the "Verifiable Credentials Specification" ...2 to 5 pages long that highlights this.

see that WG page and the charter. Much of what you're looking for is in there.

>p.s. So while I've been claiming the "specification is not implementable" (referring to the VC data model specification document), there is actually is no VC Specification document in the first place; there's just a "bunch of (good) stuff/documents". These need to be brought together ...this is a big source of the lack of cohesion in the "specification" I've been alluding to,

I think your definition and W3C's definition of what "specification" are very different. It seems what you're looking for is different than anything I've seen defined by any SDO. Happy to be shown otherwise. In this case, I'm not sure what doing here really achieves? Is this still about getting "bound and unbound credential" concepts into the spec because it feels like we're straying into an area where the discussion is more about procedure and technicalities rather than actually extending non normative and normative text. Can we stay focused on the things that are going to lead to non-normative and normative text changes here?

**Comment by mwherman2000** [comment]
In some conversations yesterday, it was suggested that some sort of _Verifiable Credentials Guide_ would be helpful. 

I'm starting work with a few people to raise this as a CCG work item (W3C Note). It will not have any impact on the current 1.1/1.2 work (at least I can't imagine it having any impact).  "More news at 11..."

**Comment by brentzundel** [comment]
It would be most helpful to re-formulate the concerns raised here as individual issues that could then be taken up by the VCWG (whether in v1.1, v1.2, or in a future version).

Raising an objection to the Verifiable Credentials Data Model 1.0  at this time doesn't have an escalation path in the W3C Process, but outlining each concern in its own issue will help future versions of the recommendation to avoid similar problems.

**Comment by brentzundel** [comment]
I also think the plan to move forward with a CCG work item is a good one.

**Comment by dlongley** [comment]
+1 to closing this once the work item is approved by the CCG.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2021-09-08-vcwg#section4-1) on 2021-09-08

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 4.1. [PRINCIPLED OBJECTION] The VC data model specification 1.0 is not implementable in the ways it appears to be intended based on the text of the current specification document (issue vc-data-model#797)

_See github issue [#797](https://github.com/w3c/vc-data-model/issues/797)._




**David Chadwick:** Michael said he is going to close it. He said this will be closed out if and when the VC Guide.... okay.  

**Brent Zundel:** Yes, which is a CCG work item  
… I'm not sure if tracking the work item there is the place for it to happen. I'm not sure how tracking it here helps.  

**David Chadwick:** The thing is the work item has not been approved yet (has it?) I think Michael is saying once it's approved, we can move on...  
… It can't mean closing it out once the guide is completed... that's going to be months, isn't it.  

**Brent Zundel:** So this doesn't get a label, and we'll look at it next time  

**David Chadwick:** I think so  

</details>

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2021-09-15-vcwg#section6-1) on 2021-09-15

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 6.1. [PRINCIPLED OBJECTION] The VC data model specification 1.0 is not implementable in the ways it appears to be intended based on the text of the current specification document (issue vc-data-model#797)

_See github issue [#797](https://github.com/w3c/vc-data-model/issues/797)._




**Kyle Den Hartog:** Short summary: this is basically something that's led to a separate CCG work item - hence why I've labeled it related content... really just a filtering mechanism so I don't have to keep having it pop up in front of me as something untriaged...  
… Intent is to have closed... given it is a CCG work item I don't think we need to formally track it.  

**Brent Zundel:** Because he has formulated it as a "principled objection", I do need to, as a chair, reach out to him... There is a little more process we have to go through before saying we are resolving this.  
… I will reach out to Michael Herman...  

**David Chadwick:** I didn't think that it was a CCG work item... I thought we have a meeting next week that it is a proposal for a new work item - and won't know if it will be accepted until next week.  

**Brent Zundel:** I agree. Hopefully another work item owner volunteers.  

**David Chadwick:** I think we've had another volunteer already.  

**Brent Zundel:** Excellent.  

**Kyle Den Hartog:** I believe that covers it for now.  

</details>

**Comment by brentzundel** [comment]
@mwherman2000 it is not clear to me if you are objecting to the publishing of the VC Data Model v1.0 Recommendation, or if you are objecting to the work we are currently engaged in to publish a revision to the Recommendation. Could you clarify?

**Comment by mwherman2000** [comment]
Closed as this is being addressed elsewhere.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2021-09-29-vcwg#section2-2) on 2021-09-29

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.2. [PRINCIPLED OBJECTION] The VC data model specification 1.0 is not implementable in the ways it appears to be intended based on the text of the current specification document (issue vc-data-model#797)

_See github issue [#797](https://github.com/w3c/vc-data-model/issues/797)._




**Wayne Chang:** We have a principled objection in 797, there has been discussion and constructive evolution since then, where do we stand with this issue?  

**Michael Herman:** Is this the one about the voting and the naming?  

**Wayne Chang:** This is the objection about VC Data Model not implementable.  

**Michael Herman:** Yes, that is resolved via the Guide, as soon as that's approved this will be closed. The VC Guide.  

**Wayne Chang:** VC Guide pending proposal in CCG.  
… Just as a word of caution, principled objections are taken extremely seriously, principled objections need to be considered ...  

**Ted Thibodeau Jr.:** Point of order, principled objections happen to CG documents, which are not REC track.  

**Wayne Chang:** It is unclear where this fits into the process of WG, VC Data Model is already published.  

**Michael Herman:** I'll follow up with wayne offline, help me understand where we are so we can get this cleaned up.  

**Brent Zundel:** I like the idea of the VC Guide, if the primary purpose of this issue is to track that work, its strange to have an item in the VCWG that tracks a CCG issue.  
… If there are additional objections that need to be resolved, we should identify what needs to happen in VCWG to see what happens.  

> *Dave Longley:* +1 for closing this issue because the resolution is to do work elsewhere, no changes here.

**Brent Zundel:** If it's a tracking issue, we should close. Otherwise, maybe keep it open.  

**David Chadwick:** This came up on CCG call yesterday, Michael needs to join CCG meeting and present, they're waiting for you to approve the work item Michael.  

> *Ryan Grant:* +1

**Ted Thibodeau Jr.:** I think that the reason this turned into a CCG work item was because Michael was not in the WG at the time and is now an invited expert. Since this has the same impact as a Note, perhaps work item should transfer back to WG, not going to change its impact in any way, cleans up what is, at present, a scrambled mess.  

**Michael Herman:** It's really simpler than all of this, the original thing was this issue and the solution was "someplace else". The issue isn't to track the work item, the VC Guide, it'll address the concern, as soon as that's approved, this issue will be closed.  
… One is not linked to the other other than issue and solution.  
… Where does the creation of the guide, jurisdiction, I don't care, I'm just writing the guide.  

**Ted Thibodeau Jr.:** Problems there include the IP agreement that apply to things, this is a weird situation in W3C, interrelationship between VCWG and CCG, this particular issue is against this WGs output, so the solution needs to be in WG output.  
… I think that was slapdash "How can we do this quickly." and it wasn't addressed.  

**Michael Herman:** I'll close the issue and we can have CCG work on it.  

> *David Chadwick:* +1

> *Ted Thibodeau Jr.:* I'm not going to fight it, but it's still more than a little strange

> *Brent Zundel:* +1 to closing

> *Dave Longley:* +1 to closing

**Wayne Chang:** I think TallTed's suggestion is pretty great, we can just continue CCG process, open up work item there, hopefully chairs there will let it through, as soon as VCWG is interested, we can continue work in this group, address concerns here.  

> *Michael Herman:* [https://github.com/w3c/vc-data-model/issues/797#issuecomment-930308901](https://github.com/w3c/vc-data-model/issues/797#issuecomment-930308901) closed

</details>

**Comment by kdenhartog** [comment]
Closing at the request of the author (based on my understanding of the meeting minutes)
